# Meta-SAIGE for Rare Variant Meta-Analysis

## Description
Meta-SAIGE is a meta-analysis tool for rare variant association studies. It is designed to combine the results of multiple cohorts and perform a meta-analysis. Meta-SAIGE is built on top of SAIGE/SAIGE-GENE+ and can be used to perform meta-analysis on the summary statistics of SAIGE and LD matrix from SAIGE-GENE+.

## Workflow Overview

![MetaSTAAR_workflow](docs/MetaSAIGE_workflow.png)

## Dependencies (built with R3.6.3)
-`SAIGE`
-`argparser`
-`data.table`
-`dplyr`
-`SKAT`
-`SPAtest`
-`Matrix`

## Input Files

-`LD matrix` : can be obtained from SAIGE `step3_LDmat.R`

-`GWAS summary` : can be obtained from SAIGE

-`Ultra-rare variant marker list` : can be obtained from SAIGE-GENE+ with `is_output_markerList_in_groupTest=TRUE` option (only for Cohort Specific collapsing)

-`Single-variant level GWAS summary for collapsed variants` : can be obtained from SAIGE-GENE+ with `--is_single_in_groupTest=TRUE` option (only for Cohort Specific collapsing)


## Command Line Usage

### GC-based Method
- `--num_cohorts` : number of cohorts
- `--chr` : chrmosome number
- `--info_file_path` : path to the marker_info.txt file generated from SAIGE 'step3_LDmat.R'. Need to specify marker_info.txt file from each and every cohort delimited by white-space (`' '`)
- `--gene_file_prefix` : prefix to the LD matrix separated by genes (also generated from SAIGE `step3_LDmat.R`) usually same as marker_info.txt file's prefix
- `--gwas_path` : path to the GWAS summary. Need to specify GWAS summary file from each and every cohort delimited by white-space (`' '`)
- `--ancestry` : ancestry indicator (ex. 1 1 1 2). Need to specify ancestry indicator from each and every cohort delimited by white-space (`' '`)
- `output_prefix`: directory for output


### Cohort-Specific Collapsing Method
- `--SingleAssociate_path` : Single variant level GWAS summary. It is needed for the GWAS summary of cohort-specifically collapsed variants (ex. test_input/cohort1/gene_based_summary/cohort1_step2_gene_res_7.txt.singleAssoc.txt)
- `--collist_path` : List of cohort specifically collapsed variants. (ex. test_input/cohort1/gene_based_summary/cohort1_step2_gene_res_7.txt.markerList.txt)
- `--groupfile` : Group file for the meta-analysis. Note that it is in a different format than SAIGE-GENE+. (It should be implemented internally to take in SAIGE-GENE+ groupfile format)
- `--annotation` : Functional annotation (ex. lof, missense_lof, missense_lof_synonymous). It has to be parsed by `_` and it has to be exactly one of three provided examples.
- `--maxMAF` : Max MAF for of the analysis
- `--Is.Hybrid` : Hybrid option uses chisq test for the collapsed variants and use ER or SPA to produce p-value.

example commands could be found in 'test_script'

## Rscript Usage

### Installation

```
library(remotes)
install_github('git@github.com:leelabsg/SAIGE_META.git')
library(MetaSAIGE)
```

### Loading Input Files
```
MetaSAIGE_InputObj <- Get_MetaSAIGE_Input(n.cohorts, chr, gwas_path, info_path, gene_file_prefix)
```

- `num_cohorts` : number of cohorts
- `chr` : chrmosome number
- `gwas_path` : a vector of paths to the SAIGE GWAS summary statistics file
- `info_file_path` : a vector of paths to the marker_info.txt file generated from SAIGE 'step3_LDmat.R'.
- `gene_file_prefix` : a vector of prefixes of the LD matrix files in the unit of genes (also generated from SAIGE `step3_LDmat.R`) usually same as marker_info.txt file's prefix

### Running Meta-Analysis
```
Run_MetaSAIGE(MetaSAIGE_InputObj, col_co, output_path, ancestry)
```
- `MetaSAIGE_InputObj` : input file objects generated by `Get_MetaSAIGE_Input` function
- `col_co` : ultra-rare variant collapsing cut-off. (10 is recommended)
- `ancestry` : ancestry indicator vector (ex. c(1, 1, 1, 2)). Need to specify ancestry indicator from each and every cohort
- `output_path` : path for the meta-analysis resutls