---
title: "Running a Meta-Analysis with MetaSAIGE"
output: 
  rmarkdown::html_vignette: default
  pdf_document: default
vignette: >
  %\VignetteIndexEntry{Running a Meta-Analysis with MetaSAIGE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


# Introduction

This vignette provides a comprehensive guide to using the `MetaSAIGE` package for conducting a meta-analysis on summary statistics from multiple cohorts. Specifically, we'll walk through the process of using the `Run_MetaSAIGE` function, detailing each step required to perform the analysis.

# Prerequisites

Before you begin the meta-analysis, ensure you have the following resources available:

- **GWAS Summary Statistics Files**: Generated by SAIGE.
- **SNP Information Files**: Corresponding to the GWAS summary statistics.
- **Gene Files**: Created by the SAIGE-GENE+ step3_LD_mat.R script.

# Running the `Run_MetaSAIGE` Function

The `Run_MetaSAIGE` function is central to performing the meta-analysis. It requires several key inputs, including the number of cohorts, chromosome number, and paths to the necessary GWAS summary statistics and SNP information files.

## Example Usage

Here is an example of how to run the `Run_MetaSAIGE` function:

```r
# Load the MetaSAIGE package
library(MetaSAIGE)

# Example of running Run_MetaSAIGE
Run_MetaSAIGE(
  n.cohorts = 2,
  chr = 1,
  gwas_path = c("gwas1.txt", "gwas2.txt"),
  info_path = c("info1.txt", "info2.txt"),
  gene_file_prefix = c("gene1_", "gene2_"),
  col_co = 10,
  output_path = "output.txt",
  ancestry = c(1, 2),
  verbose = TRUE,
  trait_type = "binary"
)
```